modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotRestoreCacheGroupProcess.java

```diff
@@ -0,0 +1,647 @@
+                        .map(v -> v.config().getGroupName() != null ? v.config().getGroupName() : v.config().getName())
+                        .collect(Collectors.toSet());
+
+                    if (!foundGrps.containsAll(cacheGrpNames)) {
+                        Set<String> missedGroups = new HashSet<>(cacheGrpNames);
+
+                        missedGroups.removeAll(foundGrps);
+
+                        fut.onDone(new IllegalArgumentException(OP_REJECT_MSG + "Cache group(s) was not found in the " +
+                            "snapshot [groups=" + missedGroups + ", snapshot=" + snpName + ']'));
+
+                        return;
+                    }
+
+                    SnapshotRestorePrepareRequest req = new SnapshotRestorePrepareRequest(
+                        UUID.randomUUID(), snpName, bltNodeIds, res.configs(), res.localNodeId());
+
+                    prepareRestoreProc.start(req.requestId(), req);
+                } catch (Throwable t) {
+                    fut.onDone(new IgniteException(OP_REJECT_MSG + t.getMessage(), t));
+                }
+            }
+        );
+
+        return new IgniteFutureImpl<>(fut);
+    }
+
+    /**
+     * Check if the cache group restore process is currently running.
+     *
+     * @return {@code True} if cache group restore process is currently running.
+     */
+    public boolean inProgress(@Nullable String cacheName) {
+        SnapshotRestoreContext opCtx0 = opCtx;
+
+        return !staleProcess(fut, opCtx0) && (cacheName == null || opCtx0.containsCache(cacheName));
+    }
+
+    /**
+     * @param fut The future of cache snapshot restore operation.
+     * @param opCtx Snapshot restore operation context.
+     * @return {@code True} if the future completed or not initiated.
+     */
+    public boolean staleProcess(IgniteInternalFuture<Void> fut, SnapshotRestoreContext opCtx) {
+        return fut.isDone() || opCtx == null;
+    }
+
+    /**
+     * Node left callback.
+     *
+     * @param leftNodeId Left node ID.
+     */
+    public void onNodeLeft(UUID leftNodeId) {
+        SnapshotRestoreContext opCtx0 = opCtx;
+
+        if (opCtx0 != null && opCtx0.nodes().contains(leftNodeId)) {
+            opCtx0.interrupt(new IgniteException(OP_REJECT_MSG +
+                "Server node(s) has left the cluster [nodeId=" + leftNodeId + ']'));
+        }
+    }
+
+    /**
+     * Abort the currently running restore procedure (if any).
+     *
+     * @param reason Interruption reason.
+     */
+    public void stop(Exception reason) {
+        SnapshotRestoreContext opCtx0 = opCtx;
+
+        if (opCtx0 != null)
+            opCtx0.interrupt(reason);
+    }
+
+    /**
+     * Ensures that a cache with the specified name does not exist locally.
+     *
+     * @param name Cache name.
+     * @throws IllegalStateException If cache with the specified name already exists.
+     */
+    private void ensureCacheAbsent(String name) throws IllegalStateException {
```

Comment:
I don't think we should declare `IllegalStateException` in the method header and it can be removed. The method is too small to check the and we can see all that is going on.